CMakeLists.txt

```diff
@@ -48,117 +60,148 @@ if(NOT WIN32)
     set(White       "${Esc}[37m")
 endif()
 
-message("-- ${BoldBlue}Radeon Performance Primitives (RPP) Version -- ${VERSION}${ColourReset}")
-message("-- ${BoldBlue}RPP install path set to -- ${CMAKE_INSTALL_PREFIX}${ColourReset}")
-message("-- ${Yellow}CMake minimum version required: ${CMAKE_MINIMUM_REQUIRED_VERSION}${ColourReset}")
-message("-- ${Green}CMake version found: ${CMAKE_VERSION}${ColourReset}")
+message("-- ${BoldBlue}AMD Radeon Performance Primitives (RPP) Version -- ${VERSION}${ColourReset}")
+message("-- ${BoldBlue}AMD RPP install path set to -- ${CMAKE_INSTALL_PREFIX}${ColourReset}")
 
-include(CheckCXXCompilerFlag)
+if(APPLE)
+  set(CMAKE_MACOSX_RPATH 1)
+  set(BACKEND "CPU")
+  message("-- ${Magenta}Apple macOS Detected -- GPU Support turned OFF${ColourReset}")
+endif()
 
-# Set defaults
-if (NOT DEFINED BACKEND)
-    SET(BACKEND "HIP")
+if(NOT DEFINED BACKEND)
+    set(BACKEND "HIP")   # set default backend to HIP
+else()
+    if("${BACKEND}" STREQUAL "OPENCL" OR "${BACKEND}" STREQUAL "OCL" OR "${BACKEND}" STREQUAL "OpenCL")
+        set(BACKEND "OCL")
+    elseif("${BACKEND}" STREQUAL "HIP" OR "${BACKEND}" STREQUAL "hip")
+        set(BACKEND "HIP")
+    elseif("${BACKEND}" STREQUAL "host" OR "${BACKEND}" STREQUAL "CPU" OR "${BACKEND}" STREQUAL "cpu")
+        set(BACKEND "CPU")
+    else()
+        message("-- ${Red}Warning: ${PROJECT_NAME} backend option unknown -- ${BACKEND}${ColourReset}")
+        message("-- ${Red}Warning: ${PROJECT_NAME} default backend will enforced${ColourReset}")
+        set(BACKEND "HIP")
+    endif()
 endif()
+message("-- ${BoldBlue}AMD RPP Backend set to -- ${BACKEND}${ColourReset}")
+
+# RPP Build Type
+if(NOT CMAKE_BUILD_TYPE)
+  set(CMAKE_BUILD_TYPE "${DEFAULT_BUILD_TYPE}" CACHE STRING "RPP Default Build Type" FORCE)
+  set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS "Debug" "Release")
+endif()
+if(CMAKE_BUILD_TYPE MATCHES Debug)
+    # -O0  -- Don't Optimize output file; -g -- generate debugging information; -Og -- Optimize for debugging experience rather than speed or size
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O0 -g -Og")
+else()
+    # -O3 -- Optimize output file; TBD:CHECK -Ofast - Optimize for speed disregarding exact standards compliance; -DNDEBUG -- turn off asserts 
```

Comment:
@rrawther @r-abishek -- do we need this -Ofast flag?

Reply:
@kiritigowda The Ofast proved to give better performance than having just O3 in the BatchPD implementations.
We could actually run the Tensor perf tests with/without Ofast to check how much difference it makes now. 
The Ofast makes a lot of reasonable NaN/+0.0/-0.0 assumptions, and utilizing associativity in arithmetics for vectorization of non-simd parts of the code.

Reply:
I think having O3 for release code is pretty safe. I haven't used Ofast. My assumption is Ofast doesn't follow standard compliance, so we shouldn't use unless required

Reply:
@r-abishek -- we can review this flag in the next phase after some experiments.

Reply:
ok